pkg/observable/channel/replay.go

```diff
@@ -0,0 +1,134 @@
+package channel
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"pocket/pkg/observable"
+)
+
+const replayNotificationTimeout = 1 * time.Second
+
+var _ observable.ReplayObservable[any] = &replayObservable[any]{}
+
+type replayObservable[V any] struct {
+	*channelObservable[V]
+	// replayBufferSize is  the number of replayBuffer to buffer so that they
+	// can be replayed to new observers.
+	replayBufferSize int
+	// replayBufferMu protects replayBuffer from concurrent access/updates.
+	replayBufferMu sync.RWMutex
+	// replayBuffer is the buffer of notifications into which new notifications
+	// will be pushed and which will be sent to new subscribers before any new
+	// notifications are sent.
+	replayBuffer []V
+}
+
+// NewReplayObservable returns a new ReplayObservable with the given replay buffer
+// replayBufferSize and the corresponding publish channel to notify it of new values.
+func NewReplayObservable[V any](
+	ctx context.Context,
+	replayBufferSize int,
+) (observable.ReplayObservable[V], chan<- V) {
+	obsvbl, publishCh := NewObservable[V]()
+	return Replay[V](ctx, replayBufferSize, obsvbl), publishCh
+}
+
+// Replay returns an observable which replays the last replayBufferSize number of
+// values published to the source observable to new observers, before publishing
+// new values.
+func Replay[V any](
```

Comment:
I don't think this function should be publically exposed. I was a bit confused if I should be using `NewReplayObservable` or `Replay` based on the structures here.

Reply:
I think there are valid use cases for both. In some scenarios you might already have an observable that you need to turn into a replay observable.